File: backend/scripts/create_owner_key.py
"""
Mint an unlimited owner API key.

Owner keys skip all credit checks (see ``backend.shared.auth.verify_api_key``) and
are meant only for the product owners themselves — there is deliberately no
HTTP endpoint for this, so it can't be triggered remotely. Run it locally
on the machine hosting the database:

    python backend/scripts/create_owner_key.py owner@example.com "Harsh"

The raw key is printed once. Store it somewhere safe (password manager,
not source control) — only its hash is kept in the database.
"""

from __future__ import annotations

import sys
from pathlib import Path

PROJECT_ROOT = Path(__file__).resolve().parent.parent.parent
if str(PROJECT_ROOT) not in sys.path:
    sys.path.insert(0, str(PROJECT_ROOT))

from backend.shared.api_keys import create_owner_key, init_api_key_tables  # noqa: E402


def main() -> None:
    if len(sys.argv) < 2:
        print("Usage: python backend/scripts/create_owner_key.py <email> [name]")
        raise SystemExit(1)

    email = sys.argv[1]
    name = sys.argv[2] if len(sys.argv) > 2 else ""

    init_api_key_tables()
    result = create_owner_key(owner_email=email, owner_name=name)

    print("Owner key created — copy it now, it will not be shown again:\n")
    print(f"  {result['api_key']}\n")
    print(f"  owner_email: {result['owner_email']}")
    print(f"  key_id:      {result['key_id']}")
    print(f"  role:        {result['role']}")


if __name__ == "__main__":
    main()

File: node-backend/src/scripts/createOwnerKey.ts
/**
 * Mint an unlimited owner API key.
 *
 * Port of `backend/scripts/create_owner_key.py`.
 *
 * Owner keys skip all credit checks (see `shared/auth.ts`) and are meant only
 * for the product owners themselves — there is deliberately no HTTP endpoint
 * for this, so it can't be triggered remotely. Run it locally on the machine
 * hosting the database:
 *
 *     npm run create-owner-key -- owner@example.com "Harsh"
 *
 * The raw key is printed once. Store it somewhere safe (password manager, not
 * source control) — only its hash is kept in the database.
 */

import { createOwnerKey } from "../shared/apiKeys.js";
import { ensureIndexes, resetClient } from "../shared/mongo.js";

async function main(): Promise<void> {
  const args = process.argv.slice(2);

  if (args.length < 1) {
    console.log("Usage: npm run create-owner-key -- <email> [name]");
    process.exit(1);
  }

  const email = args[0];
  const name = args[1] ?? "";

  // The unique index on `users.email` is what stops a second account being
  // created for the same mailbox, so it has to exist before the upsert runs.
  await ensureIndexes();

  const result = await createOwnerKey(email, name);

  console.log("Owner key created — copy it now, it will not be shown again:\n");
  console.log(`  ${result.api_key}\n`);
  console.log(`  owner_email: ${result.owner_email}`);
  console.log(`  key_id:      ${result.key_id}`);
  console.log(`  role:        ${result.role}`);

  await resetClient(null);
}

main().catch((error) => {
  console.error(error);
  process.exit(1);
});
